File: Java/src/com/leetcode/Medium/CloneGraph.java
package com.leetcode.Medium;

import java.util.*;

// 133. Clone Graph
public class CloneGraph {
    static class Node {
        public int val;
        public List<Node> neighbors;

        public Node() {
            val = 0;
            neighbors = new ArrayList<Node>();
        }

        public Node(int _val) {
            val = _val;
            neighbors = new ArrayList<Node>();
        }

        public Node(int _val, ArrayList<Node> _neighbors) {
            val = _val;
            neighbors = _neighbors;
        }
    }

    public Node cloneGraph(Node node) {
        if (node == null)
            return null;

        Queue<Node> queue = new LinkedList<>();
        Map<Node, Node> map = new HashMap<>();
        queue.add(node);
        map.put(node, new Node(node.val));

        while (!queue.isEmpty()) {
            Node n = queue.poll();
            for (Node neighbor : n.neighbors) {
                if (!map.containsKey(neighbor)) {
                    map.put(neighbor, new Node(neighbor.val));
                    queue.add(neighbor);
                }

                map.get(n).neighbors.add(map.get(neighbor));
            }
        }
        return map.get(node);
    }
}

File: JavaScript/Medium/CloneGraph.js
/**
 * 133. Clone Graph
 * // Definition for a Node.
 * function Node(val, neighbors) {
 *    this.val = val === undefined ? 0 : val;
 *    this.neighbors = neighbors === undefined ? [] : neighbors;
 * };
 */

/**
 * @param {Node} node
 * @return {Node}
 */
var cloneGraph = function (node) {
  if (node === null) return null;

  const queue = [];
  const map = new Map();
  queue.push(node);
  map.set(node, new Node(node.val));

  while (queue.length !== 0) {
    const n = queue.shift();
    for (let neighbor of n.neighbors) {
      if (!map.has(neighbor)) {
        map.set(neighbor, new Node(neighbor.val));
        queue.push(neighbor);
      }

      map.get(n).neighbors.push(map.get(neighbor));
    }
  }

  return map.get(node);
};
